ARROW-15577: [Java][Doc] Add Arrow Flight documentation (#12847)
apache/arrow: docs/source/java/flight.rst

        @Override
        // Override methods or use NoOpFlightProducer for only methods needed
    }

    Location location = Location.forGrpcInsecure("0.0.0.0", 0);
    try(
        BufferAllocator allocator = new RootAllocator();
        FlightServer server = FlightServer.builder(
                allocator,
                location,
                new TutorialFlightProducer()
        ).build();
    ){
        server.start();
        System.out.println("Server listening on port " + server.getPort());
        server.awaitTermination();
    } catch (Exception e) {
        e.printStackTrace();
    }

.. code-block:: shell

    Server listening on port 58104

Using the Flight Client
=======================

To connect to a Flight service, create a `FlightClient`_ with a location.

.. code-block:: Java

    Location location = Location.forGrpcInsecure("0.0.0.0", 58104);

    try(BufferAllocator allocator = new RootAllocator();
        FlightClient client = FlightClient.builder(allocator, location).build()){
        // ... Consume operations exposed by Flight server
    } catch (Exception e) {
        e.printStackTrace();
    }

Cancellation and Timeouts
=========================

> When making a call, clients can optionally provide `CallOptions`. This allows

Reviewer: ```suggestion
When making a call, clients can optionally provide ``CallOptions``. This allows
```

    try(BufferAllocator allocator = new RootAllocator();
        FlightClient tutorialFlightClient = FlightClient.builder(allocator, location).build()){
        try(FlightStream flightStream = flightClient.getStream(new Ticket(new byte[]{}))) {
            // ...
            flightStream.cancel("tutorial-cancel", new Exception("Testing cancellation option!"));
        }
    } catch (Exception e) {
        e.printStackTrace();
    }
    // Server
    @Override
    public void getStream(CallContext context, Ticket ticket, ServerStreamListener listener) {
        // ...
        listener.setOnCancelHandler(()->{
                    // Implement logic to handle cancellation option
                });
    }

Enabling TLS
============

TLS can be enabled when setting up a server by providing a
certificate and key pair to ``FlightServer.Builder.useTls``.

On the client side, use ``Location.forGrpcTls`` to create the Location for the client.

Enabling Authentication
=======================

.. warning:: Authentication is insecure without enabling TLS.

Handshake-based authentication can be enabled by implementing
Handlers on the client and server side (i.e.: ``ServerAuthHandler``).

> There are two options available user/password and token.

Reviewer: Both user/password and token authentication are separate - IMO let's not talk about them here, there's some work needed to document all the auth methods available in Flight and explain why you might need to use each, and I don't want to make this confusing.

Author: Thank you for the clarification. Updated.

Authentication consists of two parts: on
initial client connection, the server and client authentication
implementations can perform any negotiation needed; then, on each RPC
thereafter, the client provides a token. The client set an ID result
of user/password through ``CallHeaders`` / ``Metadata`` these IDs is
recovered by the server side to do validation. The server authentication
handler validates the token and provides the identity of the
client. This identity can be obtained from the
``CallContext.peerIdentity``.

> 

Reviewer: As evidence of what I mean, this is mixing up how different authentication schemes work. Again, the other PR already explains the basic, handshake-based authentication - let's stick to just explaining that instead of trying to explain all three schemes at once.

Author: Thank you for the clarification. Updated.

            // ...
            flightStream.cancel("tutorial-cancel", new Exception("Testing cancellation option!"));
        }
    } catch (Exception e) {
        e.printStackTrace();
    }
    // Server
    @Override
    public void getStream(CallContext context, Ticket ticket, ServerStreamListener listener) {
        // ...
        listener.setOnCancelHandler(()->{
                    // Implement logic to handle cancellation option
                });
    }

Enabling TLS
============

TLS can be enabled when setting up a server by providing a
certificate and key pair to ``FlightServer.Builder.useTls``.

On the client side, use ``Location.forGrpcTls`` to create the Location for the client.

Enabling Authentication
=======================

.. warning:: Authentication is insecure without enabling TLS.

Handshake-based authentication can be enabled by implementing
on the client side through authenticates against the Flight service
``FlightClient.authenticate``, or by authenticates with a username
and password through ``FlightClient.authenticateBasic`` and
``FlightClient.authenticateBasicToken``. And on the server side
through ``FlightServer.Builder().authHandler`` or by

> ``FlightServer.Builder().headerAuthenticator``.

Reviewer: This mixes up several different things. Can we just stick to the wording already shown? We can add the descriptions of the other auth methods later and frankly they are all not very usable.

Author: Updated

.. software distributed under the License is distributed on an
.. "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
.. KIND, either express or implied.  See the License for the
.. specific language governing permissions and limitations
.. under the License.

================
Arrow Flight RPC
================

Arrow Flight is an RPC framework for efficient transfer of Arrow data
over the network.

.. contents::

.. seealso::

   :doc:`Flight protocol documentation <../format/Flight>`
        Documentation of the Flight protocol, including how to use
        Flight conceptually.

   `Java Cookbook <https://arrow.apache.org/cookbook/java/flight.html>`_
        Recipes for using Arrow Flight in Java.

Writing a Flight Service
========================

Flight servers implement the `FlightProducer`_ interface. For convenience,
they can subclass `NoOpFlightProducer`_ instead, which offers default
implementations of all the RPC methods.

.. code-block:: Java

    public class TutorialFlightProducer implements FlightProducer {
        @Override
        // Override methods or use NoOpFlightProducer for only methods needed
    }

> Each RPC method always takes a `CallStatus` for common parameters. To indicate
> failure, raise an exception.

Reviewer: ```suggestion
Each RPC method always takes a ``CallContext`` for common parameters. To indicate
failure, pass an exception to the "listener" if present, or else raise an
exception.
```

To connect to a Flight service, call `FlightClient`_ with a location.

.. code-block:: Java

    Location location = Location.forGrpcInsecure("0.0.0.0", 58104);

    try(BufferAllocator allocator = new RootAllocator();
        FlightClient client = FlightClient.builder(allocator, location).build()){
        // ... Consume operations exposed by Flight server
    } catch (Exception e) {
        e.printStackTrace();
    }

Cancellation and Timeouts
=========================

When making a call, clients can optionally provide `CallOptions`. This allows
clients to set a timeout on calls. Also, some objects returned by client RPC calls
expose a cancel method which allows terminating a call early.

.. code-block:: Java

    Location location = Location.forGrpcInsecure("0.0.0.0", 58609);

    try(BufferAllocator allocator = new RootAllocator();
        FlightClient tutorialFlightClient = FlightClient.builder(allocator, location).build()){

        Iterator<Result> resultIterator = tutorialFlightClient.doAction(
                new Action("test-timeout"),
                CallOptions.timeout(2, TimeUnit.SECONDS)
        );
    } catch (Exception e) {
        e.printStackTrace();
    }

> On the server side, timeouts are transparent. For cancellation, the server needs to manually poll
> `setOnCancelHandler` or `isCancelled` to check if the client has cancelled the call, and if so,
> break out of any processing the server is currently doing.

Reviewer: ```suggestion
On the server side, timeouts are transparent. For cancellation, the server needs to manually poll
``setOnCancelHandler`` or ``isCancelled`` to check if the client has cancelled the call,
and if so, break out of any processing the server is currently doing.
```

.. code-block:: Java

    // Client
    Location location = Location.forGrpcInsecure("0.0.0.0", 58609);
    try(BufferAllocator allocator = new RootAllocator();
        FlightClient tutorialFlightClient = FlightClient.builder(allocator, location).build()){
        try(FlightStream flightStream = flightClient.getStream(new Ticket(new byte[]{}))) {
            // ...

>             flightStream.cancel("tutorial-cancel", new Exception("Testing cancellation opion!"));

Reviewer: ```suggestion
            flightStream.cancel("tutorial-cancel", new Exception("Testing cancellation option!"));
```

        }
    } catch (Exception e) {
        e.printStackTrace();
    }
    // Server
    @Override
    public void getStream(CallContext context, Ticket ticket, ServerStreamListener listener) {
        // ...
        listener.setOnCancelHandler(()->{
                    // Implement logic to handle cancellation option
                });
    }

Enabling TLS
============

TLS can be enabled when setting up a server by providing a

> certificate and key pair to `FlightServer.builder.useTls`.

Reviewer: ```suggestion
certificate and key pair to ``FlightServer.Builder.useTls``.
```

> On the client side, use `FlightClient.builder.trustedCertificates`.

Reviewer: ```suggestion
On the client side, use ``Location.forGrpcTls`` to create the Location for the client.
```

Enabling Authentication
=======================

.. warning:: Authentication is insecure without enabling TLS.

Handshake-based authentication can be enabled by implementing

> `ServerAuthHandler`. Authentication consists of two parts: on

Reviewer: ```suggestion
``ServerAuthHandler``. Authentication consists of two parts: on
```

initial client connection, the server and client authentication
implementations can perform any negotiation needed; then, on each RPC

> thereafter, the client provides a token. The server authentication

Reviewer: Can you look at the wording in #12815 again? This description is inaccurate

Author: Changed

Reviewer: This is still different - this doesn't describe where the client token comes from.

Author: Added where the token comes from

handler validates the token and provides the identity of the
client. This identity can be obtained from the

> `CallContext.peerIdentity`.

Reviewer: ```suggestion
``CallContext.peerIdentity``.
```

.. code-block:: Java

    // Client
    FlightClient client = FlightClient.builder().build();
    client.authenticateBasic("user", "password");
    // Server
    FlightServer server = FlightServer.builder().authHandler(new BasicServerAuthHandler(validator)).build();
    // CallHeaders printed on the Flight Server:

>     Metadata(content-type=application/grpc,user-agent=grpc-java-netty/1.44.1,auth-token-bin=bXlfdG9rZW4,grpc-accept-encoding=gzip)

Reviewer: I'd rather we omit this in favor of adding a full example to the Cookbook. Also because the server doesn't print that by default, *and* this is a different authentication method than what we just described.

Author: Deleted

Custom Middleware
=================

Servers and clients support custom middleware (or interceptors) that are called on every
request and can modify the request in a limited fashion. These can be implemented by implementing the

> `FlightServerMiddleware` and `FlightClientMiddleware` interfaces.

Reviewer: ```suggestion
``FlightServerMiddleware`` and ``FlightClientMiddleware`` interfaces.
```

> Middleware are fairly limited, but they can add headers to a request/response.

Reviewer: See #12815 which describes a little more about what middleware can do

Author: Added

.. Licensed to the Apache Software Foundation (ASF) under one
.. or more contributor license agreements.  See the NOTICE file
.. distributed with this work for additional information
.. regarding copyright ownership.  The ASF licenses this file
.. to you under the Apache License, Version 2.0 (the
.. "License"); you may not use this file except in compliance
.. with the License.  You may obtain a copy of the License at

..   http://www.apache.org/licenses/LICENSE-2.0

.. Unless required by applicable law or agreed to in writing,
.. software distributed under the License is distributed on an
.. "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
.. KIND, either express or implied.  See the License for the
.. specific language governing permissions and limitations
.. under the License.

================
Arrow Flight RPC
================

> Arrow Flight is an RPC framework for efficient transfer of Flight data

Author: Added

won't block the rest of the program. Call ``FlightServer.awaitTermination``
to block until the server stops.

.. code-block:: Java

    import org.apache.arrow.flight.Action;
    import org.apache.arrow.flight.ActionType;
    import org.apache.arrow.flight.Criteria;
    import org.apache.arrow.flight.FlightDescriptor;
    import org.apache.arrow.flight.FlightInfo;
    import org.apache.arrow.flight.FlightProducer;
    import org.apache.arrow.flight.FlightServer;
    import org.apache.arrow.flight.FlightStream;
    import org.apache.arrow.flight.Location;
    import org.apache.arrow.flight.PutResult;
    import org.apache.arrow.flight.Result;
    import org.apache.arrow.flight.SchemaResult;
    import org.apache.arrow.flight.Ticket;
    import org.apache.arrow.memory.BufferAllocator;
    import org.apache.arrow.memory.RootAllocator;
    import org.apache.arrow.util.AutoCloseables;

    class TutorialFlightProducer implements FlightProducer {
        @Override
        // Override methods or use NoOpFlightProducer for only methods needed
    }

    Location location = Location.forGrpcInsecure("0.0.0.0", 0);

    try(
        BufferAllocator allocator = new RootAllocator();

>         FlightServer server = FlightServer.builder(

Reviewer: IMO not for these examples, we don't want to use unit test helpers

.. distributed with this work for additional information
.. regarding copyright ownership.  The ASF licenses this file
.. to you under the Apache License, Version 2.0 (the
.. "License"); you may not use this file except in compliance
.. with the License.  You may obtain a copy of the License at

..   http://www.apache.org/licenses/LICENSE-2.0

.. Unless required by applicable law or agreed to in writing,
.. software distributed under the License is distributed on an
.. "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
.. KIND, either express or implied.  See the License for the
.. specific language governing permissions and limitations
.. under the License.

================
Arrow Flight RPC
================

Arrow Flight is an RPC framework for efficient transfer of Flight data
over the network.

.. contents::

.. seealso::

   :doc:`Flight protocol documentation <../format/Flight>`
        Documentation of the Flight protocol, including how to use
        Flight conceptually.

   `Java Cookbook <https://arrow.apache.org/cookbook/java/flight.html>`_
        Recipes for using Arrow Flight in Java.

Writing a Flight Service
========================

> Arrow Flight is a development framework to offer to us the building blocks
> to implement applications base on our needs.
> 
> In this journey, you need to implement how do you are planning to expose
> your business operations in base of RPC Methods and Request Patterns
> provided by Flight.
> 
> For Flight Server side operations you need to implement FlightProducer interface
> methods.

Reviewer: ```suggestion
Flight servers implement the `FlightProducer`_ interface. For convenience,
they can subclass `NoOpFlightProducer`_ instead, which offers default
implementations of all the RPC methods.
```

.. code-block:: Java

    public class TutorialFlightProducer implements FlightProducer {
        @Override
        // Override methods or use NoOpFlightProducer for only methods needed
    }

> 

Reviewer: See #12815 where we explain ServerCallContext and errors here.

Author: Added

to create and start the Flight Server. This will start the server, but won't block
the rest of the program. Call `FlightServer.awaitTermination` to block until the
server stops.

.. code-block:: Java

    import org.apache.arrow.flight.Action;
    import org.apache.arrow.flight.ActionType;
    import org.apache.arrow.flight.Criteria;
    import org.apache.arrow.flight.FlightDescriptor;
    import org.apache.arrow.flight.FlightInfo;
    import org.apache.arrow.flight.FlightProducer;
    import org.apache.arrow.flight.FlightServer;
    import org.apache.arrow.flight.FlightStream;
    import org.apache.arrow.flight.Location;
    import org.apache.arrow.flight.PutResult;
    import org.apache.arrow.flight.Result;
    import org.apache.arrow.flight.SchemaResult;
    import org.apache.arrow.flight.Ticket;
    import org.apache.arrow.memory.BufferAllocator;
    import org.apache.arrow.memory.RootAllocator;
    import org.apache.arrow.util.AutoCloseables;

    class TutorialFlightProducer implements FlightProducer {
        @Override
        // Override methods or use NoOpFlightProducer for only methods needed

>     }

Author: Deleted

    Location location = Location.forGrpcInsecure("0.0.0.0", 0);

    try(
        BufferAllocator allocator = new RootAllocator();

>         FlightServer tutorialFlightServer = FlightServer.builder(

Reviewer: ```suggestion
        FlightServer server = FlightServer.builder(
```

                allocator,
                location,
                new TutorialFlightProducer()
        ).build();
    ){
        tutorialFlightServer.start();
        System.out.println("Server listening on port " + tutorialFlightServer.getPort());

        Runtime.getRuntime().addShutdownHook(new Thread(() -> {
            try {
                System.out.println("\nExiting command...");
                AutoCloseables.close(tutorialFlightServer, allocator);
            } catch (Exception e) {
                e.printStackTrace();
            }

>         }));

Author: Deleted

        tutorialFlightServer.awaitTermination();
    } catch (Exception e) {
        e.printStackTrace();
    }

.. code-block:: shell

    Server listening on port 58104

Using the Flight Client
=======================

> To connect to a Flight service, call `FlightClient.builder` with a location.

Reviewer: Either use double-backticks or link to the javadocs.

Reviewer: The same goes for below.

Author: Added java docs link

                new Action("test-timeout"),
                CallOptions.timeout(2, TimeUnit.SECONDS)
        );
    } catch (Exception e) {
        e.printStackTrace();
    }

On the server side, timeouts are transparent. For cancellation, the server needs to manually poll
`setOnCancelHandler` or `isCancelled` to check if the client has cancelled the call, and if so,
break out of any processing the server is currently doing.

.. code-block:: Java

    import org.apache.arrow.flight.FlightClient;
    import org.apache.arrow.flight.FlightDescriptor;
    import org.apache.arrow.flight.FlightStream;
    import org.apache.arrow.flight.Location;
    import org.apache.arrow.flight.Ticket;
    import org.apache.arrow.memory.BufferAllocator;
    import org.apache.arrow.memory.RootAllocator;
    import java.nio.charset.StandardCharsets;

    // Client
    Location location = Location.forGrpcInsecure("0.0.0.0", 58609);
    try(BufferAllocator allocator = new RootAllocator();
        FlightClient tutorialFlightClient = FlightClient.builder(allocator, location).build()){
        try(FlightStream flightStream = tutorialFlightClient.getStream(new Ticket(

>                 FlightDescriptor.path("profiles").getPath().get(0).getBytes(StandardCharsets.UTF_8)))) {

Reviewer: Er, just do `new byte[]{}` or something.

Author: Added

            // ...
            flightStream.cancel("tutorial-cancel", new Exception("Testing cancellation opion!"));
        }
    } catch (Exception e) {
        e.printStackTrace();
    }
    // Server
    @Override
    public void getStream(CallContext context, Ticket ticket, ServerStreamListener listener) {
        // ...
        listener.setOnCancelHandler(()->{
                    // Implement logic to handle cancellation option
                });
    }

Enabling TLS
============

TLS can be enabled when setting up a server by providing a certificate and key pair to `FlightServer.builder.useTls`.

On the client side, use `FlightClient.builder.trustedCertificates`.

Enabling Authentication
=======================

Custom Middleware
=================

Servers and clients support custom middleware (or interceptors) that are called on every
request and can modify the request in a limited fashion. These can be implemented by implements
`FlightServerMiddleware` and `FlightClientMiddleware` interface methods.

Middleware are fairly limited, but they can add headers to a request/response.

> Example: Need to intercept Flight server headers at request and print values also for the response

Reviewer: We don't really add examples for this in the other docs, this would be better off in the cookbook

Author: Deleted

.. warning:: Authentication is insecure without enabling TLS.

Handshake-based authentication can be enabled by implementing
``ServerAuthHandler``. Authentication consists of two parts: on
initial client connection, the server and client authentication

> implementations can perform any negotiation needed; then, on each RPC

Reviewer: ```suggestion
implementations can perform any negotiation needed. The client authentication
handler then provides a token that will be attached to future calls. 
```

> thereafter, the client provides a token.

Reviewer: ```suggestion
```

The client send data to be validated through ``ClientAuthHandler.authenticate``
The server validate data received through ``ServerAuthHandler.authenticate``.
After validations, the client and server use a key value to pass/read the token
through ``Auth-Token-bin``, then the server use ``ServerAuthInterceptor``

> to validate token received from.

Reviewer: Let's not talk about how it's implemented, this is not relevant to users.

Author: Deleted
